packages/react/src/components/ComboBox/ComboBox.js

```diff
@@ -150,7 +150,7 @@ const ComboBox = (props) => {
     if (onInputChange) {
       onInputChange(inputValue);
     }
-  });
+  }, [onInputChange, inputValue]);
```

Comment:
Would it be helpful if we added a `useSavedCallback` type hook for these situations? It seems like the effect is parameterized by `inputValue` and we wouldn't want to re-run this effect when `onInputChange` happens (especially if we're passed a function that's updated each render)

`useSavedCallback` could look like:

```jsx
const savedOnInputChange = useSavedCallback(onInputChange);

// ...

useEffect(() => {
  const { current: onInputChange } = savedOnInputChange;
  if (onInputChange) {
    onInputChange(inputValue);
  }
}, [inputValue]);
```

And the implementation:

```js
function useSavedCallback(callback) {
  const ref = useRef(callback);

  useEffect(() => {
    ref.current = callback;
  });

  return ref;
}
```

Reply:
I'm not sure if I see this optimization as strictly necessary in this case but it may be worth keeping in mind

Reply:
Would `useCallback` be enough to memoize `onInputChange` and avoid unnecessarily re-running the effect?

Reply:
@emyarod I think it's something that's helpful for us to do. In this case, the effect would synchronize with the input value changing and not the handler changing (although it would need that to correctly dispatch the update).

To me, it seems similar to why they save the callback over in the [setInterval blog post](https://overreacted.io/making-setinterval-declarative-with-react-hooks/) on overreacted.

A totally extreme example of this idea would be if the parent component had a timer:

```js
function Parent() {
  const [count, setCount] = useState(0);

  useEffect(() => {
    const intervalId = window.setInterval(() => {
      setCount((count) => count + 1);
    }, 1000);
    return () => {
      window.clearInterval(intervalId);
    };
  }, []);

  return (
    <>
      <p>The current count is: {count}</p>
      <ComboBox onChange={() => { /* ... */ }} />
    </>
  );
}
```

Because we have the effect synchronizing with the callback, we end up with the effect being called repeatedly with the same value. Instead, if the callback is saved and then invoked by an effect parameterized by `inputValue` it would only run when the value changes.

Here is a quick demo of this idea: https://codesandbox.io/s/vigorous-goldwasser-rp3fe?file=/src/App.js

https://user-images.githubusercontent.com/3901764/114059480-07300c00-985a-11eb-8685-bae773bacf57.mov

Reply:
wouldn't saving the handler in a ref inside the parent function accomplish the same effect?

Reply:
Definitely. In this scenario, I'm imagining that the parent would be the person using Carbon so just was illustrating that splitting the effect out in this way would help out in the default case without them having to change how they author their code.

Reply:
something like this I assume https://codesandbox.io/s/pensive-sun-4ijti

given that the event handler is provided by the Carbon consumer, I am not sure about optimizing it for them by default (including `useCallback` usage as @tay1orjones suggested). although I think I am less hesitant about baking `useCallback` into our component if we are certain we want to optimize the handler no matter what

Reply:
I'm not opposed to optimizing it now to prevent problems down the line, though if it is an uncommon use case that the majority of consumers won't run into we can point them here (or other docs?) to instruct on how they can fix downstream in their application code.

Having this hook could be useful though, it would apply to any callback placed in a useEffect that has a dependency on one or more variables/props. I'm not sure how common that is in the codebase right now though.

Reply:
Definitely meant the counter example as the extreme case, the big thing is that in a default setup this implementation would:

- Have the effect and handler be called for non-related updates (the effect would be called because the identity changes for function declarations or arrow functions)
- Have the handler be called twice in intended situations like if the parent is using the value from `onChange` to update state (state update triggers a render that then creates a new reference for a function declaration or arrow function, this then calls the effect to be run which is the point above)

And it seems like introducing this convention would:

- Make sure the handler is only called when the value is changed and at no other time
- Make sure the handler is only called once per input change